### PR TITLE
GH-36423: [C++][Compute] Support "or" in `Expression::IsSatisfiable`

### LINKED ISSUE #36423: [C++][Compute] Can we support `or` in `Expression::IsSatisfiable`
### Describe the enhancement requested

When I go through the code, I found `Expression::IsSatisfiable` has handled case for "and" and "and_kleene" function, but "or" and "or_kleene" is not support.
If it's possible, I'd like to submit the code and testing for it

### Component(s)

C++

### PR DESCRIPTION
**Pull request by mapleFU**
### Rationale for this change

This patch supports "or" in `Expression::IsSatisfiable`, if all it child is not satisfiable, the expression itself is not satisfiable.

### What changes are included in this PR?

Support "or" and "or_kleene" in `IsSatisfiable`

### Are these changes tested?

Yes

### Are there any user-facing changes?

No

* Closes: #36423

### CODE REVIEW

**Comment by mapleFU** [comment]
I've no idea why https://github.com/apache/arrow/actions/runs/5430438414/jobs/9876174167?pr=36424 failed...

**Comment by kou** [comment]
It's #36424.

**Comment by mapleFU** [comment]
@westonpace Hi, would you mind take a look?

**Comment by pitrou** [comment]
@bkietz Do you want to take a look at this?

**Review by felipecrv — APPROVED** [review]

File: cpp/src/arrow/compute/expression.cc [hunk 1]
```diff
@@ -362,8 +362,20 @@ bool Expression::IsSatisfiable() const {
...
+      return false;
+    }
+  }
+
+  if (call->function_name == "or_kleene" || call->function_name == "or") {
+    bool all_unsatisfiable =
+        std::all_of(call->arguments.begin(), call->arguments.end(),
+                    [](const Expression& arg) { return !arg.IsSatisfiable(); });
+    if (all_unsatisfiable) {
+      return false;
```
Comment (bkietz) [inline cpp/src/arrow/compute/expression.cc:378]: IMO, using the std algorithms in this way is less readable than the for loop, please use that instead
```suggestion
    for (const Expression& arg : call->arguments) {
      if (!arg.IsSatisfiable()) return false;
    }
    return true;
  }

  if (call->function_name == "or_kleene" || call->function_name == "or") {
    for (const Expression& arg : call->arguments) {
      if (arg.IsSatisfiable()) return true;
    }
    return false;
```

Or alternatively rewrite to return the result of the std algorithms directly
```suggestion
    return std::all_of(call->arguments.begin(), call->arguments.end(),
                       [](const Expression& arg) { return arg.IsSatisfiable(); });
  }

  if (call->function_name == "or_kleene" || call->function_name == "or") {
    return std::any_of(call->arguments.begin(), call->arguments.end(),
                       [](const Expression& arg) { return arg.IsSatisfiable(); });
```

**Review by bkietz — APPROVED** [review]
LGTM, one minor nit:

File: cpp/src/arrow/compute/expression_test.cc [hunk 2]
```diff
@@ -461,6 +461,29 @@ TEST(Expression, IsSatisfiable) {
...
+           call("or", {literal(null), field_ref("bool")}),
+       }) {
+    ARROW_SCOPED_TRACE(might_true.ToString());
+    EXPECT_TRUE(Bind(might_true).IsSatisfiable());
+  }
+
+  for (const auto& never_true : {
+           // N.B. this is "or_kleene"
+           or_(literal(false), literal(null)),
+           call("or", {literal(false), literal(null)}),
```
Comment (pitrou) [inline cpp/src/arrow/compute/expression_test.cc:479]: Doesn't this test simply exercise constant folding? @bkietz What do you think?

File: cpp/src/arrow/compute/expression.cc (hunk 1 above)
Comment (mapleFU) [reply cpp/src/arrow/compute/expression.cc]: You're right. I've changed to the second one

File: cpp/src/arrow/compute/expression_test.cc (hunk 2 above)
Comment (bkietz) [reply cpp/src/arrow/compute/expression_test.cc:479]: IsSatisfiable does not (currently) do any canonicalization or simplification of the expression, so even things which could be trivially constant folded are not caught. We usually only call IsSatisfiable after simplifying by a guarantee or otherwise modifying an expression based on information available only to the caller, so we shouldn't try simplifying in IsSatisfiable since that's probably redundant and more expensive than we need.

@mapleFU could you add this to the doccomment for IsSatisfiable?

```diff
   /// Return true if this expression could evaluate to true. Will return true for any
-  /// unbound, non-boolean, or unsimplified Expressions
+  /// unbound or non-boolean Expressions. IsSatisfiable does not (currently) do any
+  /// canonicalization or simplification of the expression, so even Expressions
+  /// which are be unsatisfiable may spuriously return `true` here. This function is
+  /// intended for use in predicate pushdown where a filter expression is simplified
+  /// by a guarantee, so it assumes that trying to simplify again would be redundant.
```

Comment (mapleFU) [reply cpp/src/arrow/compute/expression_test.cc:479]: done

File: c_glib/arrow-glib/error.hpp [hunk 3]
```diff
@@ -46,7 +46,7 @@ namespace garrow {
     if (status.ok()) {
       return TRUE;
     } else {
-      std::string context = std::move(context_func());
+      std::string context = context_func();
```
Comment (pitrou) [inline c_glib/arrow-glib/error.hpp:49]: Why this change @mapleFU ?

File: cpp/src/arrow/compute/expression.h [hunk 4]
```diff
@@ -94,7 +94,11 @@ class ARROW_EXPORT Expression {
   bool IsNullLiteral() const;
 
   /// Return true if this expression could evaluate to true. Will return true for any
-  /// unbound, non-boolean, or unsimplified Expressions
+  /// unbound or non-boolean Expressions. IsSatisfiable does not (currently) do any
+  /// canonicalization or simplification of the expression, so even Expressions
+  /// which are be unsatisfiable may spuriously return `true` here. This function is
```
Comment (pitrou) [inline cpp/src/arrow/compute/expression.h:99]: Typo
```suggestion
  /// which are unsatisfiable may spuriously return `true` here. This function is
```

File: c_glib/arrow-glib/error.hpp (hunk 3 above)
Comment (mapleFU) [reply c_glib/arrow-glib/error.hpp:49]: done

**Review by pitrou — APPROVED** [review]

**Comment by pitrou** [comment]
@github-actions crossbow submit -g cpp

File: cpp/src/arrow/compute/expression.cc (hunk 1 above)
Comment (felipecrv) [reply cpp/src/arrow/compute/expression.cc]: I should have seen this. :D

File: c_glib/arrow-glib/error.hpp (hunk 3 above)
Comment (kou) [reply c_glib/arrow-glib/error.hpp]: Oh, sorry. I'll fix it by #36531.
